Add utilities which verify the offset of the virtual method on the D & C++ sides

MichaelKim20 opened the pull request:
Add utilities which verify the offset of the virtual method on the D & C++ sides

1. I added the ability to check the offset of the virtual method.
2. It checks the offset of the virtual method for any class.
3. I added a CLI that automatically generates C++ code.


Relates to #323

Geod24 commented on source/scpp/extra/DVMChecks.cpp:
```diff
@@ -0,0 +1,152 @@
+/*******************************************************************************
```
This file is generated by `cli/gen/main.d` right ?

MichaelKim20 replied:
I recorded that this file was created automatically. And I described how to make modifications.

Geod24 replied:
Great. So I don't think we should check it in the repository.

MukeunKim replied:
/Users/mukeunkim/Work/WorkSpace_BPFK/agora-node/agora/source/scpp/extra/DVMChecks_Generated.cpp

MichaelKim20 replied:
This is a video capture of running agora-gen.
[agora-gen.mov.zip](https://github.com/bpfkorea/agora/files/4418536/agora-gen.mov.zip)

MichaelKim20 replied:
log : https://gist.github.com/MukeunKim/fe9c1b848463b834260e6e5fa2453dab

Geod24 commented on source/scpd/tests/VTableTest.d:
```diff
@@ -0,0 +1,286 @@
+            }
+
+            //  If the destructor is a virtual method,
+            //  Finds two consecutive empty indexes.
```
Aren't they guaranteed to be at the beginning ?

MichaelKim20 replied:
There is no guarantee of where the order will be.

MichaelKim20 replied:
Then, do you think it is good to remove the contents related to the destructor?

MichaelKim20 replied:
I think there is no difference depending on the destructor when comparing the offset of the virtual method on C++ and D.
What do you think?

Geod24 replied:
> There is no guarantee of where the order will be.

> The order of the virtual function pointers in a virtual table is the order of declaration of the corresponding member functions in the class. If an implicitly-declared copy assignment operator, move assignment operator, or destructor is virtual, it is treated as if it were declared at the end of the class, in that order. (Implicitly-declared assignment operators may be virtual if a base class declares a virtual assignment operator taking a reference to a derived class type.)

https://itanium-cxx-abi.github.io/cxx-abi/abi.html#vtable-components

Geod24 commented:
Side note: When comparing pointers, use `is` or `!is`

MichaelKim20 commented:
@Geod24 Ready for review.

Geod24 commented on source/scpd/tests/VTableTypes.d:
```diff
@@ -0,0 +1,35 @@
+    abstract void vfunc2();
+}
+
+public const string[] VTableCheckClasses =
```
```suggestion
public immutable VTableCheckClasses =
```

Geod24 commented on source/scpd/tests/VTableTest.d:
```diff
@@ -16,6 +16,13 @@ module scpd.tests.VTableTest;
+
+version (unittest)
+{
+    extern (C++) long getVMOffsetTestA(const char* name);
```
```suggestion
    extern(C++) long getVMOffsetTestA (const char* name);
```

Geod24 reviewed (commented):
I still don't understand the point of the first few commits.
It starts with all the checks being hardcoded, and gradually hardcode less and less. But the point of the issue is to have the generator, so those intermediate steps, while valuable for you in your search for a solution, shouldn't be in the final version.

Geod24 commented on source/agora/cli/gen/main.d:
```diff
@@ -0,0 +1,200 @@
+
+*******************************************************************************/
+
+module agora.cli.gen.main;
```
`gen` is a bit too generic of a name. Something like `check-vtable` would be more accurate.

Geod24 commented on source/agora/cli/gen/main.d:
```diff
@@ -0,0 +1,200 @@
+            file.writeln(`{`);
+
+            method_idx = 1;
+            static foreach (member; __traits(allMembers, %1$s))
```
How are overloads handled ?

MichaelKim20 replied:
I don't know why this place needs overloads.
Why do you think it is necessary?
The parameter and return values are not used here.
I knew it was decided not to take care of the override.

MichaelKim20 replied:
There is also `getVMOffset...`, what about this?

MichaelKim20 replied:
``` D
/// It checks the offset of virtual on C++ and D for any class.
unittest
{
    //  Offset of  C++ side
    long[string] offset_cpp;

    //  Offset of D side
    long[string] offset_d;
    long val_d;

    static foreach (e; VTableCheckClasses)
    {
        offset_cpp.clear();
        offset_d.clear();
        val_d = 0;

        mixin(
        q{
            static foreach (member; __traits(allMembers, %1$s))
            {
                mixin(
                q{
                    static if (__traits(isVirtualMethod, %1$s.%2$s) && (`%2$s` != `__xdtor`))
                    {
                        static if (`%2$s` == `__dtor`)
                        {
                            // for D side
                            val_d += 2;
                        }
                        else
                        {
                            // for D side
                            offset_d[`%2$s`] = val_d++;

                            // for C++ side
                            offset_cpp[`%2$s`] = getVMOffset%1$s(`%2$s`);
                        }
                    }
                }.format(`%1$s`, member));
            }

            assert(offset_d == offset_cpp, format(`%%s : %%s`, offset_d, offset_cpp));

        }.format(e, `%2$s`));
    }
}
```

Geod24 commented on source/scpp/extra/DVMChecks.cpp:
```diff
@@ -0,0 +1,320 @@
+}
+
+///  Uses to test SCPDriver.
+class SCPDriverForTest
```
We shouldn't need to declare this extra type on the C++ side. Can't we just get the offset on the D side ?

MichaelKim20 replied:
> We shouldn't need to declare this extra type on the C++ side. Can't we just get the offset on the D side ?

I wonder if this type of data is possible.
map<string, int>

Geod24 reviewed (commented):
Comments:
- First few commits are steps in development and shouldn't be here;
- Last commit shouldn't be in either: let's not checkout the generated file. Instead I'll add a check to the CI.
- Overloads are completely ignored, so if upstream introduces an overload, this will break;
- The method of declaring the class on the C++ side is quite brittle too. Why can't we get that offset from the D side and pass it to `checkVMOffsetSCPDriver` ?

MichaelKim20 commented:
@Geod24 
I think I know a little about what you want.
I removed most of the C++ side code, leaving only part of it.


I'm done with the modifications.
The automatically generated code is here.
https://gist.github.com/MukeunKim/796b82581a76b16ce9e3929844b76406

Geod24 commented on source/agora/cli/checkvtable/main.d:
```diff
@@ -0,0 +1,193 @@
+
+*******************************************************************************/
+
+module agora.checkvtable.gen.main;
```
That should be `agora.cli.checkvtable.main`

Geod24 commented on source/agora/cli/checkvtable/main.d:
```diff
@@ -0,0 +1,193 @@
+/*******************************************************************************
+
+    Entry point for the agora-gen
```
This also. And since it's a single-file program, please add a description of what it does and why.

MichaelKim20 commented:
@Geod24 
Ready for review

I left only the code below on C++.

``` C++
#include "DVMChecks.h"

int checkVMOffset (const char * classname, const char* offsets)
{
    return 0;
}
```
I added one feature to the agora-checkvtable

The automatically generated code is here.
https://gist.github.com/MukeunKim/53f0e9d9183c3dde92df4b99d74e22bd

MichaelKim20 commented:
In C++
``` C++
class TestA
{
public:
    virtual void foo1(int a) {};
    virtual void foo2(int a) {};
};
```

In D
``` D

extern(C++) abstract class TestA
{
    abstract void foo2 (int a);
    abstract void foo1 (int a);
}

extern(C++) class BonTestA : TestA
{
    public override void foo2 (int a) { writeln("Hello World 2"); }
    public override void foo1 (int a) { writeln("Hello World 1"); }
}

pragma(msg, BonTestA.foo1.mangleof);
pragma(msg, BonTestA.foo2.mangleof);

mixin("extern(C) void ", BonTestA.foo1.mangleof, "(void* _this, int);");
mixin("alias callMe1 = ", BonTestA.foo1.mangleof, ";");

mixin("extern(C) void ", BonTestA.foo2.mangleof, "(void* _this, int);");
mixin("alias callMe2 = ", BonTestA.foo2.mangleof, ";");

void main ()
{
    writefln("&callMe1 =       %s", &callMe1);
    writefln("&callMe2 =       %s", &callMe2);
    writefln("&BonTestA.foo1 = %s", &BonTestA.foo1);
    writefln("&BonTestA.foo2 = %s", &BonTestA.foo2);

    auto m = new BonTestA();

    writefln("m.__vptr[0] = %s", m.__vptr[0]);
    writefln("m.__vptr[1] = %s", m.__vptr[1]);
}
```
```
&callMe1 =       10143648C
&callMe2 =       101436464
&BonTestA.foo1 = 10143648C
&BonTestA.foo2 = 101436464
m.__vptr[0] = 101436464
m.__vptr[1] = 10143648C
```

Both callMe1 and callMe2 have the addresses of virtual methods on the D side.

AndrejMitrovic commented on source/scpd/tests/VTableTest.d:
```diff
@@ -0,0 +1,75 @@
+/// In case the virtual methods are in the incorrect order.
+unittest
+{
+    // abstract class, this is in the incorrect order.
```
What do you mean, incorrect order? The vtable will be based on the order of the methods as defined in the **base** class, not the inherited class. It doesn't matter what the order is in the derived class.

In your `BonTestB` derived class you can reorder the methods and your tests will still pass. But these are the sort of tests that belong in the compiler's test-suite, not Agora.

MichaelKim20 replied:
Yes, I agree with you.
This is not to solve the issue.
But this will allow us to understand the issue correctly.
In conclusion, I don't think it's necessary. I'll remove it.

AndrejMitrovic replied:
> But this will allow us to understand the issue correctly.

We could write some documentation about it, if it's unclear. I think we have an issue for that. Here: #560

AndrejMitrovic commented on source/scpp/extra/DVMTestTypes.cpp:
```diff
@@ -0,0 +1,20 @@
+/*******************************************************************************
```
This seems to belong in the next commit, it's not used in this commit.

AndrejMitrovic commented on source/scpp/extra/DVMTestTypes.h:
```diff
@@ -0,0 +1,30 @@
+/*******************************************************************************
```
It belongs in the next commit.

AndrejMitrovic commented on source/scpd/tests/VTableTest.d:
```diff
@@ -73,3 +78,44 @@ unittest
+    import std.string;
+
+    long[string] offset_d;
+    long val_d;
```
What is "val_d"? Can you add a comment or rename it to something more obvious?

AndrejMitrovic commented on source/scpd/tests/VTableTest.d:
```diff
@@ -73,3 +78,44 @@ unittest
+                    static if (__traits(isVirtualMethod, %1$s.%2$s) && (`%2$s` != `__xdtor`))
+                    {
+                        static if (`%2$s` == `__dtor`)
+                            val_d += 2;
```
Why += 2?

MichaelKim20 replied:
The destructor always occupies two pointers on the vtable.
So I added two.

AndrejMitrovic replied:
Ah yes, I think on the D side there are two destructors. One is `__dtor` and one is `__xdtor`.

MichaelKim20 replied:
The offsets of TestA are 
TestA._dtor : 0
TestA.vfunc1 : 1
TestA.vfunc2 : 2
TestA.__xdtor : 3
If `__dtor` and `__xdtor` come out one after another, we can add 1 more, but since that's not the case, we have to add 2.

MichaelKim20 replied:
``` C++
class A
{
public:
    virtual ~A() {};
    virtual void vfunc1(){};
    virtual void vfunc2(){};
};

class B
{
public:
    virtual void vfunc1(){};
    virtual void vfunc2(){};
    virtual ~B() {};
};

int CPPTest1 ()
{
    auto pAf1 = &A::vfunc1;
    printf("A.vfunc1 : %d\n", (long)reinterpret_cast<void *&>(pAf1) / sizeof(long));
    auto pAf2 = &A::vfunc2;
    printf("A.vfunc2 : %d\n", (long)reinterpret_cast<void *&>(pAf2) / sizeof(long));
    auto pBf1 = &B::vfunc1;
    printf("B.vfunc1 : %d\n", (long)reinterpret_cast<void *&>(pBf1) / sizeof(long));
    auto pBf2 = &B::vfunc2;
    printf("B.vfunc2 : %d\n", (long)reinterpret_cast<void *&>(pBf2) / sizeof(long));
}
```
Result is 
A.vfunc1 : 2
A.vfunc2 : 3
B.vfunc1 : 0
B.vfunc2 : 1

In A, 0 and 1 are destructors, and in B, 2 and 3 are destructors.

AndrejMitrovic commented on source/scpp/extra/DVMChecks.cpp:
```diff
@@ -0,0 +1,19 @@
+
+int checkVMOffset (const char* classname, const char* offsets)
+{
+    return 0;
```
What is the purpose of this function? It just returns 0.

MichaelKim20 replied:
@Geod24 gave me this comment.
https://github.com/bpfkorea/agora/pull/711#pullrequestreview-386915944

I just want to keep that pattern at unittest.
And I thought that A would be automatically generated in CI and the test would proceed.
I couldn't implement overloaded methods with C++ source.
So I had no choice but to make this decision.

AndrejMitrovic commented on source/agora/cli/checkvtable/main.d:
```diff
@@ -0,0 +1,222 @@
+/*******************************************************************************
+
+    Entry point for the agora-checkvtable
+        It generate C++ code for chacking the offset of the vtable.
+        The path to the file it creates is `source/scpp/extra/DVMChecks.cpp`.
+
+        The use of this is as follows.
+        agora-checkvtable source/scpp/extra/DVMChecks.cpp
```
```suggestion
    It generates C++ code for checking the offset of the vtable.
    The path to the file it creates is `source/scpp/extra/DVMChecks.cpp`.

    The use of this is as follows.
    agora-checkvtable source/scpp/extra/DVMChecks.cpp
```

AndrejMitrovic commented on source/agora/cli/checkvtable/main.d:
```diff
@@ -0,0 +1,222 @@
+/*******************************************************************************
+
+    Entry point for the agora-checkvtable
+        It generate C++ code for chacking the offset of the vtable.
+        The path to the file it creates is `source/scpp/extra/DVMChecks.cpp`.
+
+        The use of this is as follows.
+        agora-checkvtable source/scpp/extra/DVMChecks.cpp
```
It should not generate temporary files in the source tree. It should generate them in a temporary directory, verify the offsets work, and then exit with 0. We should not commit these autogenerated files into the repository. 

Because any time any of the SCP classes change, we will have to remember to commit new files into the repository. We want an **automated** testing process.

MichaelKim20 replied:
It's just an example.
The fIle `source/scpp/extra/DVMChecks.cpp` is not an automatically generated file.

MichaelKim20 replied:
For automation, the source must be generated before the pre-builder.
I think the processing of this will have to make another CI.

AndrejMitrovic replied:
Ok so is your plan to add this CI step in this PR, or in another PR?

MichaelKim20 replied:
I didn't have any specific plans for CI, but with your help, I thought a lot about it.

I will remove the file below from the current PR.
source/scpd/tests/VTableTest.d
source/scpp/extra/DMMCheks.cpp

I think CI needs the following functions.
1. C++ Source Generator (D): I will use CLI checkvtable
2. Prebuilder (C++): I will use source/scpp/build.d.
3. A small project for small checks.
This will include source/scpd/tests/VTableTest.d.

And I will make CI in another PR.

MichaelKim20 replied:
I implemented this as follows.
A new feature has been added to the CLI agora-checkvtable.
Previously, there was only the ability to create C++source.
This can be done with the command:
agora-checkvtable generation outfilename
And I added the function to check the offset of the virtual method.
This can be done with the command:
agora-checkvtable check
And I added a test code that can be used by CI
ci/check_vtable_test.d

AndrejMitrovic commented on source/agora/cli/checkvtable/main.d:
```diff
@@ -0,0 +1,222 @@
+
+*******************************************************************************/
+
+int main (string[] args)
```
The entire body of this function should be another function which either returns a `string` or writes to a `buffer`. That way, you can unittest the function and see that it generates the right code.

AndrejMitrovic replied:
And then in `main()` you would just write that buffer to a file.

MichaelKim20 replied:
@AndrejMitrovic That's a good idea. Thank you.

MichaelKim20 replied:
I tried this.
There is an error message that the computer time cannot be accessed. 
I could not change to another string array instead of VTableCheckClasses.

AndrejMitrovic replied:
> There is an error message that the computer time cannot be accessed.

Could you show the full error message and also the code which is causing the issue?

MichaelKim20 replied:
I fixed the source as below, and when I tested it, an error occurred.

``` D
/// This is agora-checkvtable sub-function for generating C++ source code
int generate (string[] args, ref string[] outputs)
{
    if (args.length < 3)
    {
        writeln("usage: agora-gen command [output source file]");
        return 1;
    }

    auto filename = args[2];
    writecode(VTableCheckClasses, outputs);
    File file = File(filename, "w+");

    foreach(ref line; outputs)
        file.writeln(line);

    file.close();

    writefln("Source code was generated with the file name %s.", filename);
    return 0;
}

/// Writes C++ source code to outputs
void writecode (immutable(string[]) types, ref string[] outputs)
{
    writeHeader(outputs);

    static foreach (e; types)
    {
        mixin(
        q{
            writeMethodHeader(outputs, `%1$s`, `getVMOffset%1$s`);
            static foreach (idx, member; __traits(allMembers, %1$s))
            {
                mixin(
                q{
                    static if (__traits(isVirtualMethod, %1$s.%2$s) && (`%2$s` != `__dtor`) && (`%2$s` != `__xdtor`))
                    {
                        writeMethodContents(outputs, `%1$s`, `%2$s`);
                    }
                }.format(`%1$s`, member));
            }

            writeMethodFooter(outputs);
            writeCheckFunction(outputs, `%1$s`);
            outputs ~= ``;

        }.format(e, `%2$s`));
    }

    outputs ~= `int checkVMOffset (const char* classname, const char* offsets)`;
    outputs ~= `{`;

    static foreach (e; types)
    {
        mixin(
        q{
            outputs ~= `    if (strcmp(classname, "%1$s") == 0)`;
            outputs ~= `    {`;
            outputs ~= `        return checkVMOffset%1$s(offsets);`;
            outputs ~= `    }`;
        }.format(e));
    }

    outputs ~= `    return -1;`;
    outputs ~= `}`;
}
```

source/agora/cli/checkvtable/generate.d(55,24): Error: variable types cannot be read at compile time
source/agora/cli/checkvtable/generate.d(81,24): Error: variable types cannot be read at compile time

MichaelKim20 replied:
I solved this by using CTFE.

AndrejMitrovic commented on source/agora/cli/checkvtable/main.d:
```diff
@@ -0,0 +1,222 @@
+}
+
+/// Writes the header of the source file
+void writeHeader(File file)
```
It should write to a buffer or an output sink (just like the Formatter and the Serializer)

AndrejMitrovic replied:
Also I have the same comment for all the other functions which write to a file.

AndrejMitrovic commented on source/agora/cli/checkvtable/main.d:
```diff
@@ -0,0 +1,222 @@
+/// Writes the contents of the method
+void writeMethodContents(File file, bool first, string entry, string method)
+{
+    if (first)
```
You could simplify this by removing the need to use `first`. So instead of generating this:

```C++
if (strcmp(name, "foo") == 0)
{
    auto pf = class::foo;
    return  ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
}
else if (strcmp(name, "bar") == 0)
{
    auto pf = class::bar;
    return  ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
}
```

You could remove the `first` parameter and generate this:


```C++
if (strcmp(name, "foo") == 0)
{
    auto pf = class::foo;
    return ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
}

if (strcmp(name, "bar") == 0)
{
    auto pf = class::bar;
    return ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
}
```

The end result is the same (short-circuiting evaluation) as using `if/else if`, because there is a `return` statement in each block.

Geod24 commented:
Ping on the review and the commits: It looks like everything should be squashed to one commit ?

MichaelKim20 commented:
> Ping on the review and the commits: It looks like everything should be squashed to one commit ?

@Geod24 Ready for review

Geod24 commented:
Has the rest of the review been addressed ? It doesn't seem so.

AndrejMitrovic commented:
> @Geod24: Last commit shouldn't be in either: let's not checkout the generated file. Instead I'll add a check to the CI.

So then this PR here needs some support work from @Geod24?

@MukeunKim is there anything else blocking you?

MukeunKim commented:
> > @Geod24: Last commit shouldn't be in either: let's not checkout the generated file. Instead I'll add a check to the CI.
> 
> So then this PR here needs some support work from @Geod24?
> 
> @MukeunKim is there anything else blocking you?

If we've confirmed some of the current structure, I'll have to make CI next time.
I think the current structure needs to be changed a little bit when we create CI.
Maybe we don't need a CLI, maybe we need a small project.
Therefore, we should first listen to @Geod24's plan and proceed with the development accordingly.

MichaelKim20 commented:
> > @Geod24: Last commit shouldn't be in either: let's not checkout the generated file. Instead I'll add a check to the CI.
> 
> So then this PR here needs some support work from @Geod24?
> 
> @MukeunKim is there anything else blocking you?

@AndrejMitrovic 
If we've confirmed some of the current structure, I'll have to make CI next time.
I think the current structure needs to be changed a little bit when we create CI.
Maybe we don't need a CLI, maybe we need a small project.
Therefore, we should first listen to @Geod24's plan and proceed with the development accordingly.

MichaelKim20 commented:
@AndrejMitrovic I think I applied everything I could to your proposal.
I couldn't find a way for the next one.
Is there a good way?
https://github.com/bpfkorea/agora/pull/711#discussion_r407955740

Geod24 commented on dub.json:
```diff
@@ -119,6 +124,29 @@
+            "mainSourceFile": "source/agora/cli/checkvtable/main.d",
+            "sourceFiles-posix": [
+                "source/scpp/build/DSizeChecks.o",
+                "source/scpp/build/DLayoutChecks.o",
```
Is it needed ?

Geod24 commented on dub.json:
```diff
@@ -106,11 +107,15 @@
-                "source/scpp/build/DLayoutChecks.o"
+                "source/scpp/build/DLayoutChecks.o",
+                "source/scpp/build/DVMTestTypes.o",
+                "source/scpp/build/DVMChecks.o"
```
Is it needed ?

Geod24 commented on dub.json:
```diff
@@ -106,11 +107,15 @@
-                "source/scpp/build/DLayoutChecks.obj"
+                "source/scpp/build/DLayoutChecks.obj",
+                "source/scpp/build/DVMTestTypes.o",
+                "source/scpp/build/DVMChecks.o"
```
That's obviously wrong, so probably not needed

Geod24 commented on dub.json:
```diff
@@ -119,6 +124,29 @@
+                "source/scpp/build/DSizeChecks.obj",
+                "source/scpp/build/DLayoutChecks.obj",
+                "source/scpp/build/DVMTestTypes.o",
+                "source/scpp/build/DVMChecks.o"
```
Not needed

Geod24 commented on source/agora/cli/checkvtable/main.d:
```diff
@@ -0,0 +1,58 @@
+
+/// Workaround for issue likely related to dub #225,
+/// expects a main() function and invokes it after unittesting.
+version (unittest) void main () { } else:
```
This code should never be compiled with `-unittest` so this is not needed.

MichaelKim20 replied:
Yes

Geod24 commented on ci/check_vtable_test.d:
```diff
@@ -0,0 +1,54 @@
+
+private int main (string[] args)
+{
+    // If the user pass `nobuild` as first argument, skip build,
+    // Build `agora-checkvtable`
+    if (args.length < 2 || args[1] != "nobuild")
+        runCmd(BuildBinary);
```
Let's remove this

MichaelKim20 replied:
Okay

Geod24 commented on dub.json:
```diff
@@ -119,6 +122,27 @@
+            ]
+        },
+        {
+            "name": "checkvtable",
```
As discussed this needs to be split in 2 configurations

MichaelKim20 replied:
I fixed it

Geod24 commented on source/agora/cli/checkvtable/check.d:
```diff
@@ -0,0 +1,74 @@
+import std.stdio;
+import std.string;
+
+version (Windows)
```
Add a message here

MichaelKim20 replied:
I fixed it as follows.
```
version (Windows) { pragma(msg, "This does not support Windows."); } else:
```

MichaelKim20 committed:
Add CLI to check the offsets of a virtual method

The agora-checkvtable generate C++ source code.
The agora-checkvtable check the offsets of a virtual method on the D & C++ sides.
Add test code for CI

Co-Authored-By: Mathias LANG <pro.mathias.lang@gmail.com>

MichaelKim20 commented:
@Geod24 Ready for review.

AndrejMitrovic commented on source/agora/cli/checkvtable/generate.d:
```diff
@@ -0,0 +1,318 @@
+    string[] outputs;
+    auto filename = args[1];
+    writecode(outputs);
+    File file = File(filename, "w+");
```
This can be just `"w"` I think?

MichaelKim20 replied:
Good catch.

AndrejMitrovic commented on source/agora/cli/checkvtable/generate.d:
```diff
@@ -0,0 +1,318 @@
+import std.format;
+import std.stdio;
+
+version (Windows) { pragma(msg, "This does not support Windows."); } else:
```
What was wrong with checking the vtable on Windows?

MichaelKim20 replied:
> What was wrong with checking the vtable on Windows?

I tested 'MS Visual C++' on the following site.
https://rextester.com/l/cpp_online_compiler_visual

``` C++
#include <iostream>

class A
{
public:
    virtual ~A() {};
    virtual void vfunc1(){};
    virtual void vfunc2(){};
};

int main ()
{
    auto pAf1 = &A::vfunc1;
    std::cout << pAf1;
    
    return 0;
}

```
It worked abnormally in MS Visual C++.

MichaelKim20 replied:
If the offset of vtable's virtual methods is proven to be the same in `gcc`, then of course in Windows, we no longer have to make unnecessary efforts.
The test will also be run only on CI.

AndrejMitrovic commented on source/agora/cli/checkvtable/types.d:
```diff
@@ -0,0 +1,34 @@
+
+import std.meta;
+
+extern (C++) public class TestA
```
We don't need this class anymore right? It was only used during development. We only care about the SCPDriver class.

MichaelKim20 replied:
> We don't need this class anymore right? It was only used during development. We only care about the SCPDriver class.

I needed a function that could handle several things using 'foreach statement'. I needed more than one class to test this.

MichaelKim20 replied:
I removed TestA. In the third unittest(source/agora/cli/checkvtable/generate.d), I tested one or more classes.

AndrejMitrovic reviewed (commented):
Just two small comments.

MichaelKim20 commented:
Ping @AndrejMitrovic @Geod24

AndrejMitrovic commented:
This PR adds the vtable check generator, but it never uses it? Where is the call to `./ci/check_vtable_test.d`?

MichaelKim20 commented:
> This PR adds the vtable check generator, but it never uses it? Where is the call to `./ci/check_vtable_test.d`?

@AndrejMitrovic I added `./ci/check_vtable_test.d` to `.circleci/config.yml`.

AndrejMitrovic commented:
I tested it locally and it works.

AndrejMitrovic commented on .circleci/config.yml:
```diff
@@ -41,6 +41,13 @@ jobs:
+          name: Check virtual method offset
+          command: |
+            source $HOME/dlang/ldc-1.20.1/activate
+            ci/check_vtable_test.d
```
By the way, we could have added this check into `run.sh` instead? Is there a reason why it was moved to CircleCI?

MichaelKim20 replied:
https://github.com/bpfkorea/agora/pull/711#pullrequestreview-386915944
I just didn't want to have a big impact on Agora's development environment.

Geod24 commented on .circleci/config.yml:
```diff
@@ -41,6 +41,13 @@ jobs:
+      - run:
+          name: Check virtual method offset
+          command: |
+            source $HOME/dlang/ldc-1.20.1/activate
```
```suggestion
            source $HOME/dlang/ldc-1.21.0/activate
```

Geod24 commented on .circleci/config.yml:
```diff
@@ -41,6 +41,13 @@ jobs:
+          command: |
+            source $HOME/dlang/ldc-1.20.1/activate
+            ci/check_vtable_test.d
+            # Work around druntime setting the permissions to 600..
```
```suggestion
```

Geod24 reviewed (approved):
Just a minor fix, otherwise LGTM

Geod24 commented:
You also need to rebase on `v0.x.x` it seems

Geod24 commented on .circleci/config.yml:
```diff
@@ -28,7 +28,7 @@ jobs:
           command: |
             mkdir -p $HOME/dlang && wget https://dlang.org/install.sh -O $HOME/dlang/install.sh
             chmod +x $HOME/dlang/install.sh
-            $HOME/dlang/install.sh install ldc-1.20.1
```
https://github.com/bpfkorea/agora/commit/c79a5103c995c969af0a6c0d792d2799fd8497d3
